### DISCUSSION TITLE
Problem with Cart API When Using Function APIs

### DISCUSSION THREAD

**Discussion by axome-vchangea** [Shopify/function-examples#482] [Feedback: _Development and Deployment_]
Hello,

I am encountering an issue with the Cart API when a discount is applied via the Product Discount API to the cart. It appears that, as soon as a discount is active, some functionalities of the API, specifically those related to change.js for modifying the cart based on the line key, no longer work correctly.

Reproduction Scenario:

Add an item to the cart via the Cart API.
Apply a discount to the cart via Shopify Function.
Attempt to modify this item using its line key via change.js.

Expected Behavior:

The cart should allow the item to be modified despite the application of the discount.

Observed Behavior:

Once the discount is applied, attempts to modify the item in the cart using its line key via change.js fail with a 400 error. However, using the line index can patch the problem.

![image](https://github.com/Shopify/function-examples/assets/63097729/c7053ac5-4150-4096-a347-e882da984fd1)

![image](https://github.com/Shopify/function-examples/assets/63097729/57c303d7-fd14-4af5-9975-28044f674317)

**Comment by API2CartOfficial** [comment]
Here's a structured approach to potentially resolve or further diagnose the problem:

Initial Diagnosis

- Verify API Documentation: Ensure your usage aligns with the latest Shopify API documentation, especially Cart API and Discount APIs. Shopify frequently updates its APIs, so it is crucial to confirm that your code adheres to the latest standards.

- Error Code Analysis: A 400 error usually indicates a bad request. Reviewing the exact error message returned by the API can provide insights into what might be going wrong, such as issues with the request syntax, deprecated fields, or validation errors.

Steps to Troubleshoot

1. Isolate the Issue: Confirm if the issue is specifically triggered by the application of the discount or if it could be related to how items are being updated in the cart:
2. Try modifying the cart item without applying any discounts to see if change.js works as expected.
3. Apply a discount, but try modifying a different aspect of the cart or a different item to see if the problem persists across different scenarios.
4. Check the Payload: Inspect the request payload you are sending through change.js. Ensure the line item key used in the request is correct and hasn’t been altered by the discount application process.
5. Use API Versioning: If you are not already specifying an API version in your requests, consider setting it to ensure stability in how the APIs behave. Sometimes, behaviors change with new API versions but stabilizing the version can provide consistent results.
6. Line Index vs. Line Key: Since you mentioned that using the line index patches the problem, there could be a specific issue with how the line keys are interpreted by Shopify's backend after a discount is applied. This might warrant deeper investigation: Log the line key and line index before and after the discount is applied. Compare these logs to identify any changes or discrepancies.

If all else fails, reach out to Shopify Developer Support. 

Also, if you are experiencing continuous problems with Shopify API, I recommend trying the API2Cart solution. It provides a unified API for [integration with Shopify](https://api2cart.com/supported-platforms/shopify-integration/?utm_source=github&utm_medium=referral&utm_campaign=shopifyintegrationn.kov) and other platforms.
